tests/topotests/lib/common_config.py

```diff
@@ -1619,60 +1625,60 @@ def interface_status(tgen, topo, input_dict):
     return True
 
 
-def retry(attempts=3, wait=2, return_is_str=True, initial_wait=0, return_is_dict=False):
+def retry(retry_timeout, initial_wait=0, expected=True):
     """
-    Retries function execution, if return is an errormsg or exception
-
-    * `attempts`: Number of attempts to make
-    * `wait`: Number of seconds to wait between each attempt
-    * `return_is_str`: Return val is an errormsg in case of failure
-    * `initial_wait`: Sleeps for this much seconds before executing function
+    Fixture: Retries function while it's return value is an errormsg (str), False, or it raises an exception.
 
+    * `retry_timeout`: Retry for at least this many seconds; after waiting initial_wait seconds
+    * `initial_wait`: Sleeps for this many seconds before first executing function
+    * `expected`: if False then the return logic is inverted, except for exceptions,
+                      (i.e., a False or errmsg (str) function return ends the retry loop,
+                      and returns that False or str value)
     """
 
     def _retry(func):
         @wraps(func)
         def func_retry(*args, **kwargs):
-            _wait = kwargs.pop("wait", wait)
-            _attempts = kwargs.pop("attempts", attempts)
-            _attempts = int(_attempts)
-            if _attempts < 0:
-                raise ValueError("attempts must be 0 or greater")
+            _retry_sleep = 2
+
+            _retry_timeout = kwargs.pop("retry_timeout", retry_timeout)
+            _expected = kwargs.pop("expected", expected)
+            _initial_wait = kwargs.pop("initial_wait", initial_wait)
+
+            retry_until = datetime.now() + timedelta(seconds=_retry_timeout + _initial_wait)
 
             if initial_wait > 0:
                 logger.info("Waiting for [%s]s as initial delay", initial_wait)
                 sleep(initial_wait)
 
-            _return_is_str = kwargs.pop("return_is_str", return_is_str)
-            _return_is_dict = kwargs.pop("return_is_str", return_is_dict)
-            _expected = kwargs.setdefault("expected", True)
-            kwargs.pop("expected")
-            for i in range(1, _attempts + 1):
+            invert_logic = not _expected
+            attempts = 0
+            while True:
+                seconds_left = (retry_until - datetime.now()).total_seconds()
+                attempts += 1
                 try:
                     ret = func(*args, **kwargs)
                     logger.debug("Function returned %s", ret)
-                    if _return_is_str and isinstance(ret, bool) and _expected:
-                        return ret
-                    if (
-                        isinstance(ret, str) or isinstance(ret, unicode)
-                    ) and _expected is False:
-                        return ret
-                    if _return_is_dict and isinstance(ret, dict):
```

Comment:
Hi @choppsv1 we should consider return value as dict also, few APIs returns dictionary as result.

Reply:
The code does, and handles this as modified.

Reply:
I see `ret is False or is_string(ret)` : will that recognize a dict ?

Reply:
Correct. if it's not negative it's positive. :)

Reply:
IOW any result other than False, an error string, or an exception is considered successful

Reply:
Also for the purposes of inverting the logic an exception is not considered successful. To that end I could easily be convinced that an exception should never be retried as well, as the API of the functions this fixture is applied to do not raise exceptions to report a normal "False" condition. IOW Exceptions truly are exceptional.